File: Project/C++/Project1/Eshop.cpp
#include "Eshop.h"
#include<iostream>
#include"Pen.h"
#include"Paper.h"
#include"Notebook.h"
#include"Pencil.h"
Eshop::Eshop()
{
}
Eshop::Eshop(std::string name, Owner* owner)
{
	this->name = name;
	this->owner = owner;
	
}

void Eshop::addItem(Item* item)
{
	//---Duplicate check---
	bool itemExists = false;
	for (int i = 0; i < itemsList.size(); i++) {
		if (itemsList[i]->equals(*item)) {
			itemExists = true;
		}
	}
	//---------------------
	if (!itemExists) {
		itemsList.push_back(item);
	}
	else {
		throw(std::invalid_argument("Item already exists"));
	}
}

void Eshop::addBuyer(Buyer* buyer)
{
	//---Duplicate check---
	bool buyerExists = false;
	for (int i = 0; i < buyersList.size(); i++) {
		if (buyersList[i]->equals(*buyer)) {
			buyerExists = true;
		}
	}
	//---------------------
	if (!buyerExists) {
		buyersList.push_back(buyer);
		
	}
	else {
		throw(std::invalid_argument("Buyer already exists"));
	}
}

void Eshop::removeItem(Item* item)
{
	//---Existance check---
	bool itemExists = false;
	for (int i = 0; i < itemsList.size(); i++) {
		if (itemsList[i]->equals(*item)) {
			itemExists = true;
		}
	}
	//---------------------
	if (itemExists) {
		int j = 0;
		for (int i = 0; i < itemsList.size(); i++) {
			if (itemsList[i]->equals(*item)) {
				itemsList.erase(itemsList.begin()+i);
			}
		}
	}
	else {
		throw(std::invalid_argument("Item does not exist"));
	}
}

void Eshop::removeBuyer(Buyer* buyer)
{
	//---Existance check---
	bool buyerExists = false;
	for (int i = 0; i < buyersList.size(); i++) {
		if (buyersList[i]->equals(*buyer)) {
			buyerExists = true;
		}
	}
	//---------------------
	if (buyerExists) {
		buyer->cart.clearCart();
		for (int i = 0; i < buyersList.size(); i++) {
			if (buyersList[i]->equals(*buyer)) {
				buyersList.erase(buyersList.begin() + i);
			}
		}
	}
	else {
		throw(std::invalid_argument("Buyer does not exist"));
	}
}

Item* Eshop::getItemByld(int id)
{
	bool itemExists = false;
	for (int i = 0; i < itemsList.size(); i++) {
		if (id == itemsList[i]->getId()) {
			itemExists = true;
			return itemsList[i];
		}
	}
	if (!itemExists) {
		throw(std::invalid_argument("Item does not exist"));
	}
}

Buyer* Eshop::getBuyerByEmail(std::string email)
{
	bool buyerExists = false;
	for (int i = 0; i < buyersList.size(); i++) {
		if (email.compare(buyersList[i]->getEmail()) == 0) {
			buyerExists = true;
			return buyersList[i];
		}
	}
	if (!buyerExists) {
		throw(std::invalid_argument("Buyer does not exist"));
	}
}

void Eshop::updateItemStock(Item* item, int stock)
{
	bool itemExists = false;
	for (int i = 0; i < itemsList.size(); i++) {
		if (itemsList[i]->equals(*item)) {
			itemExists = true;
			itemsList[i]->setStock(stock);
		}
	}
	if (!itemExists) {
		throw(std::invalid_argument("Item does not exist"));
	}
}

void Eshop::showCategories()
{
	std::cout << "1:Pen(" << Pen::ammount << ")" << "\n2:Notebook(" << Notebook::ammount << ")" << "\n3:Paper("<< Paper::ammount << ")" <<"\n4:Pencil(" << Pencil::ammount << ")";
}

void Eshop::showProductsInCategory(std::string cat)
{
	for (int i = 0; i < itemsList.size(); i++) {
		if (cat.compare(itemsList[i]->getObjType()) == 0) {
			std::cout << itemsList[i]->getBasicInfo() << "\n";
		}
	}
}

void Eshop::showProduct(Item item)
{
	item.toString();
}

void Eshop::checkStatus()
{
	for (int i = 0; i < buyersList.size(); i++) {
		std::cout << i+1 << ":Name: " << buyersList[i]->getName() << ". Email: " << buyersList[i]->getEmail() << ". Points: " << buyersList[i]->getBuyerPts() << ". Category: ";
		if (buyersList[i]->getBuyerCategory() == 0) {
			std::cout << "Bronze.\n";
		}
		else if (buyersList[i]->getBuyerCategory() == 1) {
			std::cout << "Silver.\n";
		}
		else {
			std::cout << "Gold.\n";
		}
	}
}

User* Eshop::checkBuyer(std::string email)
{
	bool buyerExists = false;
	for (int i = 0; i < buyersList.size(); i++) {
		if (buyersList[i]->getEmail().compare(email) == 0) {
			return buyersList[i];
			buyerExists = true;
			break;
		}
	}
	if (!buyerExists) {
		if (owner->getEmail().compare(email) == 0) {
			return owner;
		}
		else {
			throw(std::invalid_argument("Buyer does not exist"));
		}
	}
}

File: Project/Java/src/eshop/EShop.java
package eshop;

import java.util.ArrayList;
import java.util.List;

public class EShop {
    private String name;
    private Owner owner;
    private List<Buyer> buyersList = new ArrayList<Buyer>();
    private List<Item> itemsList = new ArrayList<Item>();

    EShop(Owner owner){
        this.owner = owner;
    }

    public User checkBuyer(String email)throws BuyerNotFoundException{
        boolean pass = false;
        int saveI = 0;
        for(int i = 0; i<buyersList.size(); i++){
            if((buyersList.get(i).getEmail()).equals(email)){
                pass = true;
                saveI = i;
            }
        }

        if(owner.getEmail().equals(email) && !pass){
            return owner;
        }

        if(pass){
            return buyersList.get(saveI);
        }else{
            throw new BuyerNotFoundException();
        }
    }

    public void addItem(Item item){
        if(!itemsList.contains(item)){
            itemsList.add(item);
        }
    }

    public Item getItemByld(int id) throws ItemNotFoundException{
        for(int i = 0; i < itemsList.size(); i++){
            if(itemsList.get(i).getId() == id){
                return itemsList.get(i);
            }
        }
        throw new ItemNotFoundException();
    }

    public void removeItem(Item item) throws ItemNotFoundException{

        itemsList.remove(item);
    }

    public void addBuyer(Buyer buyer)throws BuyerAlreadyExistsException{
        if(buyersList.contains(buyer)){
            throw new BuyerAlreadyExistsException();
        }else{
            buyersList.add(buyer);
        }
    }

    public void removeBuyer(Buyer buyer) throws BuyerNotFoundException{
        buyer.cart.clearCart();
        buyersList.remove(buyer);
    }

    public void updateItemStock(Item item, int stock){
        item.setStock(stock);

    }

    public void showCategories(){
        System.out.println("1:Pen("+Pen.getCount()+")" + "\n2:Notebook("+Notebook.getCount() + ")" + "\n3:Paper("+Paper.getCount()+")" + "\n4:Pencil("+Pencil.getCount()+")");
    }

    public List<Item> showProductsInCategory(Class cat){
        List<Item> products = new ArrayList<Item>();
       for(int i = 0; i<itemsList.size(); i++){
           if(itemsList.get(i).getClass() == cat){
               products.add(itemsList.get(i));
               System.out.println(i+1 + ":"+itemsList.get(i).toString());
           }
       }
       return products;
    }

    public void showProduct(Item item) throws ItemNotFoundException{
        System.out.println(item.toString());
        throw new ItemNotFoundException();
    }

    public void checkStatus(){
        for(int i = 0; i<buyersList.size(); i++){
            System.out.print("Name: " + buyersList.get(i).getName() + " Email: " + buyersList.get(i).getEmail() + " Points: " + buyersList.get(i).getBuyerPts() + " Category: ");
            if(buyersList.get(i).getbuyerCategory() == 0){
                System.out.println("Bronze");
            }else if(buyersList.get(i).getbuyerCategory() == 1){
                System.out.println("Silver");
            }else{
                System.out.println("Gold");
            }
        }
    }
}
